readme.markdown
ArgyrisPack
===========
Overview
--------
ArgyrisPack is a library for doing things with the Argyris-5 finite element, a
C1 finite element with 21 degrees of freedom. The Argyris-5 element spans the
space of 5th degree polynomials on each element and has O(h^4) convergence in
the H2 norm. The main goals of ArgyrisPack are numerical correctness,
performance, and portability. ArgyrisPack is mostly written in C and can be called
from C, Matlab, Julia, or Python.

ArgyrisPack is available under the BSD-3 License. See the file `License.txt` for
more details.

Design & Usage Philosophy
-------------------------
ArgyrisPack assumes that it is faster to work with precomputed values of
reference basis functions. Therefore it is up to the user to set up appropriate
tables of reference data (function values, derivative values, presumably second
derivative values) and to use these when setting up the typical per-element
matricies.

Since performance and portability are both goals, the C code does not offer
much abstraction; one passes in pointers to arrays and the functions fill the
arrays appropriately. The non-C interfaces take care of wrapping output values
and checking bounds instead.

Compilation Details
-------------------
ArgyrisPack is written in C99 and may be compiled to use either row-major or
column-major storage.  Macros (contained in `order_logic.h`) ensure that the C
code does 'the right things' (correct BLAS calls, correct order traversal, etc)
in both cases. This is done by passing either `-DUSE_ROW_MAJOR` or
`-DUSE_COL_MAJOR` to the compiler. In particular, with the makefile, compile the
library with

    make STORAGE_ORDER=USE_ROW_MAJOR

to pass one symbol or another. To compile the mex-files, run

    make

at the MATLAB prompt. MATLAB (to the best of my knowledge) assumes MEX files on
*nix platforms will use the gcc, but you may need to tweak your local MEX
settings to compile the binaries correctly. Julia just needs `libargyris_pack.so`
in the current path to work. Python needs for the .so to be in the working path
as well.

ArgyrisPack currently targets GNU/Linux and OS X. GCC 4.2 and later should work;
please let us know if this is not the case. In principle, there is nothing
preventing a port to other operating systems, but all the provided makefiles
assume a *nix environment.

What can ArgyrisPack do?
------------------------
ArgyrisPack consists of four levels of C and Python functions/classes:
* Level 1: evaluation of the Argyris basis functions (and derivatives) at
  arbitrary points on the reference triangle. These are the functions beginning
  with `ref_`.
* Level 2: evaluation of the required coordinate transformations (Dominguez and
  Sayas' C, B, b, and Theta (Th) matricies) and evaluation of Argyris basis
  functions on arbitrary triangles. These are the functions beginning with
  `physical_`.
* Level 3: evaluation of a few common bilinear forms (the classic stiffness and
  mass matricies as well as the 'biharmonic' matrix resulting from
  discretization of the biharmonic operator). These functions begin with
  `matrix_`.
* Level 4: mesh generation. Argyris elements have 21 nodes (5 on each corner
  and one at the midpoint of each triangle edge). ArgyrisPack contains mesh
  parsing and creation classes for a variety of textual representations of
  meshes.

There are a few additional files; we wrote a 'multiply by a diagonal matrix'
routine, wrappers to make `dgemm` work with row or column order, as well as a
symbolic (`symbolic.py` and `symbolic.m`) version of the Argyris element to
verify the numerical computations.

Running the Tests
-----------------
The tests are currently implemented in Python. They require SAGE to be installed
and used as a library. This is because the tests work by interpolating the
Argyris polynomials on a physical element and then comparing the symbolic
results with the numerical results. To run the results from SAGE, go to the root
directory of ArgyrisPack and execute

    import ap.test; ap.test.run()

The meshing software has tests, but these are not linked to the numerical tests
at the moment.

Authors
-------
* Erich Foster, Virginia Tech
* Traian Iliescu, Virginia Tech
* David Wells, Virginia Tech

Relevant Papers
---------------
ArgyrisPack is a new implementation of older ideas. We are greatly indebted
to those who have done the necessary mathematical research for making this
library possible.

* Dominguez and Sayas: A simple Matlab implementation of the Argyris element,
  2006
* Argyris, Fried and Scharpf: The TUBA Family of Plate Elements for the Matrix
  Displacement Method, 1968.


history.txt
July 13 2013 : V2.12 : Included new projection code. Cleaned up Python code with
                       pep8 and more pylint.
June 16 2013 : V2.11 : Fix a bug in projections.py. Cleaned up Python code with
                       help from pylint.
June 8 2013  : V2.10 : Various bugfixes regarding the 2.0 release; Python
                       interface and tests implemented. Changed the directory
                       from 'numerical' to 'numeric'.
May 16 2013  : V2.00 : Post-thesis release of new meshing materials and renaming
                       of local to reference and global to physical.
Oct 25 2012  : V1.06 : bugfix release; fixed a problem with the no-projection
                       case.
Oct 25 2012  : V1.05 : bugfix release; include a workaround to delete nodes that
                       are not in any triangle.
Oct 19 2012  : V1.04 : bugfix release; change runMESH.py to use the included
                       rectangle.mesh.
Oct 18 2012  : V1.03 : bugfix release; slight change in interface for
                       construction of Argyris meshes. Includes sample mesh.
Sep 20 2012  : V1.02 : bugfix release; also supports DGEMM_WRAPPER_TN macro.
Sep 15 2012  : V1.01 : Removed unnecessary .geo and the test.c file
Sep 14 2012  : V1.00 : Finished symbolic implementation; release to public under BSD-3


symbolic.m
argyris = containers.Map('KeyType', 'double', 'ValueType', 'any');
argyrisdx = containers.Map('KeyType', 'double', 'ValueType', 'any');
argyrisdy = containers.Map('KeyType', 'double', 'ValueType', 'any');
argyrisdxx = containers.Map('KeyType', 'double', 'ValueType', 'any');
argyrisdxy = containers.Map('KeyType', 'double', 'ValueType', 'any');
argyrisdyy = containers.Map('KeyType', 'double', 'ValueType', 'any');

syms x y

% functions for the triangle xs = [1/2, 3/2, 1], ys = [1, 1, 3/2]
argyris(1) = -6*x^5 - 60*x^3*y^2 - 120*x^2*y^3 - 30*x*y^4 + 30*x^4 + 24*y^5 + 120*x^3*y + 600*x^2*y^2 + 360*x*y^3 - 115*x^3 - 150*y^4 - 840*x^2*y - 1185*x*y^2 + 405*x^2 + 290*y^3 + 1410*x*y - 120*y^2 - 4575/8*x - 150*y + 875/8;
argyris(2) = 6*x^5 + 60*x^3*y^2 - 120*x^2*y^3 + 30*x*y^4 - 30*x^4 + 24*y^5 - 120*x^3*y + 240*x^2*y^2 + 120*x*y^3 + 115*x^3 - 210*y^4 - 120*x^2*y - 495*x*y^2 - 45*x^2 + 530*y^3 + 510*x*y - 570*y^2 - 1185/8*x + 270*y - 371/8;
argyris(3) = 240*x^2*y^3 - 48*y^5 - 840*x^2*y^2 - 480*x*y^3 + 360*y^4 + 960*x^2*y + 1680*x*y^2 - 360*x^2 - 820*y^3 - 1920*x*y + 690*y^2 + 720*x - 120*y - 62;
argyris(4) = -3*x^5 - 28*x^3*y^2 - 46*x^2*y^3 + 7*x*y^4 + 31/2*x^4 + 22*y^5 + 56*x^3*y + 248*x^2*y^2 + 50*x*y^3 - 115/2*x^3 - 315/2*y^4 - 358*x^2*y - 321*x*y^2 + 723/4*x^2 + 809/2*y^3 + 464*x*y - 448*y^2 - 3335/16*x + 405/2*y - 725/32;
argyris(5) = -8*x^4*y - 2*x^3*y^2 + 2*x^2*y^3 - 22*x*y^4 + 8*x^4 - 18*y^5 + 38*x^3*y - 9*x^2*y^2 + 100*x*y^3 - 36*x^3 + 139*y^4 - 38*x^2*y - 311/2*x*y^2 + 45*x^2 - 807/2*y^3 + 255/2*x*y + 2205/4*y^2 - 50*x - 725/2*y + 375/4;
argyris(6) = -3*x^5 - 28*x^3*y^2 + 46*x^2*y^3 + 7*x*y^4 + 29/2*x^4 - 22*y^5 + 56*x^3*y - 80*x^2*y^2 - 134*x*y^3 - 107/2*x^3 + 287/2*y^4 + 22*x^2*y + 335*x*y^2 + 129/4*x^2 - 641/2*y^3 - 296*x*y + 322*y^2 + 1289/16*x - 293/2*y + 785/32;
argyris(7) = -8*x^4*y + 2*x^3*y^2 + 2*x^2*y^3 + 22*x*y^4 + 8*x^4 - 18*y^5 + 26*x^3*y - 21*x^2*y^2 - 108*x*y^3 - 28*x^3 + 95*y^4 - 2*x^2*y + 431/2*x*y^2 + 21*x^2 - 391/2*y^3 - 351/2*x*y + 753/4*y^2 + 46*x - 167/2*y + 55/4;
argyris(8) = -68*x^3*y^2 + 20*x*y^4 + 136*x^3*y + 204*x^2*y^2 - 116*x*y^3 - 68*x^3 - 20*y^4 - 408*x^2*y + 41*x*y^2 + 204*x^2 + 116*y^3 + 186*x*y - 177*y^2 - 131*x + 86*y - 5;
argyris(9) = -28*x^2*y^3 - 20*y^5 + 98*x^2*y^2 + 56*x*y^3 + 114*y^4 - 112*x^2*y - 196*x*y^2 + 42*x^2 - 279*y^3 + 224*x*y + 727/2*y^2 - 84*x - 246*y + 135/2;
argyris(10) = -1/2*x^5 - 5/2*x^3*y^2 - 9/2*x^2*y^3 + x*y^4 + 11/4*x^4 + 5/2*y^5 + 5*x^3*y + 95/4*x^2*y^2 + 7/2*x*y^3 - 33/4*x^3 - 18*y^4 - 34*x^2*y - 231/8*x*y^2 + 163/8*x^2 + 375/8*y^3 + 173/4*x*y - 859/16*y^2 - 685/32*x + 105/4*y - 225/64;
argyris(11) = -2*x^4*y - 3*x^3*y^2 - x^2*y^3 - x*y^4 + 2*x^4 - y^5 + 15*x^3*y + 23/2*x^2*y^2 + 6*x*y^3 - 12*x^3 + 15/2*y^4 - 69/2*x^2*y - 77/4*x*y^2 + 24*x^2 - 87/4*y^3 + 137/4*x*y + 257/8*y^2 - 20*x - 105/4*y + 75/8;
argyris(12) = 1/2*x^3*y^2 - 1/2*x^2*y^3 - 5/2*x*y^4 - 3/2*y^5 - x^3*y + 1/4*x^2*y^2 + 27/2*x*y^3 + 1/2*x^3 + 49/4*y^4 + x^2*y - 201/8*x*y^2 - 3/4*x^2 - 305/8*y^3 + 79/4*x*y + 903/16*y^2 - 45/8*x - 40*y + 175/16;
argyris(13) = 1/2*x^5 + 5/2*x^3*y^2 - 9/2*x^2*y^3 - x*y^4 - 9/4*x^4 + 5/2*y^5 - 5*x^3*y + 35/4*x^2*y^2 + 29/2*x*y^3 + 25/4*x^3 - 16*y^4 - 4*x^2*y - 289/8*x*y^2 - 25/8*x^2 + 287/8*y^3 + 131/4*x*y - 583/16*y^2 - 291/32*x + 67/4*y - 181/64;
argyris(14) = 2*x^4*y - 3*x^3*y^2 + x^2*y^3 - x*y^4 - 2*x^4 + y^5 - x^3*y + 13/2*x^2*y^2 + 2*x*y^3 + 4*x^3 - 11/2*y^4 - 15/2*x^2*y - 37/4*x*y^2 + 55/4*y^3 + 49/4*x*y - 125/8*y^2 - 4*x + 31/4*y - 11/8;
argyris(15) = -1/2*x^3*y^2 - 1/2*x^2*y^3 + 5/2*x*y^4 - 3/2*y^5 + x^3*y + 13/4*x^2*y^2 - 23/2*x*y^3 - 1/2*x^3 + 29/4*y^4 - 5*x^2*y + 145/8*x*y^2 + 9/4*x^2 - 105/8*y^3 - 47/4*x*y + 179/16*y^2 + 21/8*x - 9/2*y + 11/16;
argyris(16) = 9*x^2*y^3 - 5*y^5 - 59/2*x^2*y^2 - 18*x*y^3 + 65/2*y^4 + 32*x^2*y + 59*x*y^2 - 23/2*x^2 - 299/4*y^3 - 64*x*y + 619/8*y^2 + 23*x - 71/2*y + 43/8;
argyris(17) = 6*x^3*y^2 + 2*x*y^4 - 12*x^3*y - 18*x^2*y^2 - 6*x*y^3 + 6*x^3 - 2*y^4 + 36*x^2*y + 45/2*x*y^2 - 18*x^2 + 6*y^3 - 35*x*y - 21/2*y^2 + 33/2*x + 11*y - 9/2;
argyris(18) = x^2*y^3 + 3*y^5 - 7/2*x^2*y^2 - 2*x*y^3 - 35/2*y^4 + 4*x^2*y + 7*x*y^2 - 3/2*x^2 + 165/4*y^3 - 8*x*y - 393/8*y^2 + 3*x + 59/2*y - 57/8;
argyris(19) = 16*x^4*y - 32*x^2*y^3 - 16*x^4 + 16*y^5 - 64*x^3*y + 128*x^2*y^2 + 64*x*y^3 + 64*x^3 - 112*y^4 - 72*x^2*y - 256*x*y^2 - 24*x^2 + 280*y^3 + 272*x*y - 304*y^2 - 80*x + 145*y - 25;
argyris(20) = -64*sqrt(1/2)*x^3*y^2 - 64*sqrt(1/2)*x^2*y^3 + 64*sqrt(1/2)*x*y^4 + 64*sqrt(1/2)*y^5 + 128*sqrt(1/2)*x^3*y + 416*sqrt(1/2)*x^2*y^2 - 192*sqrt(1/2)*x*y^3 - 480*sqrt(1/2)*y^4 - 64*sqrt(1/2)*x^3 - 640*sqrt(1/2)*x^2*y - 48*sqrt(1/2)*x*y^2 + 1328*sqrt(1/2)*y^3 + 288*sqrt(1/2)*x^2 + 416*sqrt(1/2)*x*y - 1672*sqrt(1/2)*y^2 - 240*sqrt(1/2)*x + 960*sqrt(1/2)*y - 200*sqrt(1/2);
argyris(21) = -64*sqrt(1/2)*x^3*y^2 + 64*sqrt(1/2)*x^2*y^3 + 64*sqrt(1/2)*x*y^4 ...
    - 64*sqrt(1/2)*y^5 + 128*sqrt(1/2)*x^3*y - 32*sqrt(1/2)*x^2*y^2 - 448*sqrt(1/2)*x*y^3 + 352*sqrt(1/2)*y^4 - 64*sqrt(1/2)*x^3 - 128*sqrt(1/2)*x^2*y + 848*sqrt(1/2)*x*y^2 - 688*sqrt(1/2)*y^3 + 96*sqrt(1/2)*x^2 - 608*sqrt(1/2)*x*y + 616*sqrt(1/2)*y^2 + 144*sqrt(1/2)*x - 256*sqrt(1/2)*y + 40*sqrt(1/2);

for i=1:21
    argyrisdx(i) = diff(argyris(i), x);
    argyrisdy(i) = diff(argyris(i), y);
    argyrisdxx(i) = diff(argyris(i), x, 2);
    argyrisdxy(i) = diff(diff(argyris(i), y), x);
    argyrisdyy(i) = diff(argyris(i), y, 2);
end

% generate comparison points.
xs = rand(50,1);
ys = (1 - xs).*rand(50,1);

[C B b Th] = apGlobalMapsMex([1/2, 3/2, 1], [1, 1, 3/2]);

functionRef            = apLocalFunctionsMex(xs, ys);
[dxRef dyRef]          = apLocalGradientsMex(xs, ys);
[dxxRef dxyRef dyyRef] = apLocalHessiansMex(xs, ys);

values = apGlobalFunctionsMex(C, functionRef);
[dx dy] = apGlobalGradientsMex(C, B, dxRef, dyRef);
[dxx dxy dyy] = apGlobalHessiansMex(C, Th, dxxRef, dxyRef, dyyRef);

% map the comparison points to global coordinates.
globalPoints = B*[xs';ys'];
globalxs = globalPoints(1,:) + b(1);
globalys = globalPoints(2,:) + b(2);

valuesSymbolic = zeros(size(values));
dxSymbolic = zeros(size(dx));
dySymbolic = zeros(size(dy));
dxxSymbolic = zeros(size(dxx));
dxySymbolic = zeros(size(dxy));
dyySymbolic = zeros(size(dyy));

for i=1:21
    for j=1:size(valuesSymbolic, 2)
        valuesSymbolic(i,j) = subs(subs(argyris(i), x, globalxs(j)), y, globalys(j));
        dxSymbolic(i,j)     = subs(subs(argyrisdx(i), x, globalxs(j)), y, globalys(j));
        dySymbolic(i,j)     = subs(subs(argyrisdy(i), x, globalxs(j)), y, globalys(j));
        dxxSymbolic(i,j)    = subs(subs(argyrisdxx(i), x, globalxs(j)), y, globalys(j));
        dxySymbolic(i,j)    = subs(subs(argyrisdxy(i), x, globalxs(j)), y, globalys(j));
        dyySymbolic(i,j)    = subs(subs(argyrisdyy(i), x, globalxs(j)), y, globalys(j));
    end
end

disp('Error Results:')
disp(['Max Error in function values : ' max(max(abs(values - ...
                                                  valuesSymbolic)))]);
disp(['Max Error in dx              : ' max(max(abs(dx - dxSymbolic)))]);
disp(['Max Error in dy              : ' max(max(abs(dy - dySymbolic)))]);
disp(['Max Error in dxx             : ' max(max(abs(dxx - dxxSymbolic)))]);
disp(['Max Error in dxy             : ' max(max(abs(dxy - dxySymbolic)))]);
disp(['Max Error in dyy             : ' max(max(abs(dyy - dyySymbolic)))]);